src/ray/core_worker/lessor_picker.cc

```diff
@@ -0,0 +1,58 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/core_worker/lessor_picker.h"
+
+namespace ray {
+
+absl::optional<rpc::Address> LessorPicker::GetBestNodeForTask(
+    const TaskSpecification &spec) {
+  if (auto node_id = GetBestNodeIdForTask(spec)) {
+    return node_addr_factory_(node_id.value());
+  }
+  return absl::nullopt;
+}
+
+absl::optional<NodeID> LessorPicker::GetBestNodeIdForTask(const TaskSpecification &spec) {
+  return GetBestNodeIdForObjects(spec.GetDependencyIds());
+}
+
+/// Criteria for "best" node: The node with the most object bytes (from object_ids) local.
+absl::optional<NodeID> LessorPicker::GetBestNodeIdForObjects(
+    const std::vector<ObjectID> &object_ids) {
+  // Number of object bytes (from object_ids) that a given node has local.
+  absl::flat_hash_map<NodeID, uint64_t> bytes_local_table;
+  uint64_t max_bytes = 0;
+  absl::optional<NodeID> max_bytes_node;
+  // Finds the node with the maximum number of object bytes local.
+  for (const ObjectID &object_id : object_ids) {
+    if (auto locality_data = locality_data_provider_->GetLocalityData(object_id)) {
+      for (const NodeID &node_id : locality_data->nodes_containing_object) {
+        auto &bytes = bytes_local_table[node_id];
```

Comment:
Isn't this crashing? I don't see anywhere that we emplace node id to this table?

Reply:
Unless I'm mistaken, `absl::flat_hash_map<Key, T>::operator[]` should follow the `std::unordered_map<Key, T>::operator[]` semantics of [inserting a new entry when the key doesn't exist](https://en.cppreference.com/w/cpp/container/unordered_map/operator_at), returning a reference to that inserted value, value-initialized to 0 in our case.

>Returns a reference to the value that is mapped to a key equivalent to key, performing an insertion if such key does not already exist.

>Inserts a value_type object constructed in-place from `std::piecewise_construct, std::forward_as_tuple(key), std::tuple<>()` if the key does not exist... When the default allocator is used, this results in the key being copy constructed from key and the mapped value being value-initialized.

Am I understanding those semantics correctly? And I'll make sure to confirm that this is the case when writing unit tests for `LessorPicker`.